docs/internal/rfc/009rfc-transition-explorer.md

```diff
@@ -0,0 +1,459 @@
+ * representation as a `TlaEx` or see an associated operator name (if any)?  */
+type Transition
+
+/** An execution is an alternating sequence of states and transitions,
+ *  terminating in a state with no transition.
+ *
+ * It is a high-level representation of the `EncodedExecution` maintained by the
+ * transition executor).
+ *
+ * E.g., an execution from `s1` to `sn` with transitions `ti` for each `i` in `1..n-1`:
+ * 
+ *     List((s1, Some(t1)), (s2, Some(t2)), ..., (sn, None))
+ */
+type Execution = List[(State, Option[Transition])]
+
+trait UninitializedModel = {
+  val spec: TlaModule
+  val transitions: Set(Transition)
+}
+
+trait InitializedModel extends UninitializedModel {
+  val constInitPrimed: Map[String, TlaEx]
+}
+
+/** An abstract representation of the `CheckerInput`
+ *
+ * A `Model` includes all static data representing a model.
+ * 
+ * An `UninitializedModel` is missing the information that would be needed in
+ * order to actually explore its symbolic states (such as the initial values of
+ * its constants).
+ *
+ * An `InitializedModel` has all data needed to explore its symbolic states. */
+type Model = Either[UninitializedModel, InitializedModel]
+
+/**  The type of errors that can result from failures when loading a spec. */
+type LoadErr
+
+/**  The type of errors that can result from failures to assert a constraint. */
+type AssertErr
+
+/**  The type of errors that can result from checking satisfiability of a frame. */
+type SatError
+
+/**  Maintains session and connection information */
+type Connection
+
+trait TransEx {
+
+  /** Used internally */
+  private def connection(): Connection
+
+  /**  Create a transition explorer
+   *
+   * Establishes the connection and a running session
+   *
+   * The channel is managed by the gRPC framework. */
+  def apply(channel: Channel): TransEx
+
+
+  /** Reset the state of the explorer
+   *
+   * Returns the explorer to the same state as when the currently loaded model
+   * was freshly loaded. Used to restart exploration from a clean slate.
+   *
+   * [TRANS-EX.1::LOAD.1]
+   */
+  def reset(): Unit
+
+  /** Load a model for exploration
+   *
+   * If a model is already loaded, it will be replaced and the state of the exploration
+   * [[reset]].
+   *
+   * [TRANS-EX.1::QCHECK.1]
+   * [TRANS-EX.1::LOAD.1]
+   * [TRANS-EX.1::SBMC.1::TRANSITIONS.1]
+   * 
+   * @param spec the root TLA+ module 
+   * @param aux auxiliary modules that may be required by the root module
+   * @return `Left(LoadErr)` if parsing or loading the model from `spec` goes
+   *          wrong, or `Right(UninitializedModel)` if the model is loaded successfully.
+   */
+  def loadModel(spec: String, aux: List[String] = List()): Either[LoadErr, UninitializedModel]
+
+  /** Initialize the constants of a model
+   *
+   * This will always also reset an exploration back to its initial frame. */
+  def initializeConstants(constants: Map[String, TlaEx]): Either[AssertErr, InitializedModel]
```

Comment:
This part probably requires special note:

- @Kukovec previously asked that it be impossible to load a module that was in an invalid state, suggesting instead that the constants be initialized when the model is loaded.
- @konnov subsequently requested that we *not* initialize the constants when the model is created, and that we instead just have a single method to assert constraints, and use that also for initializing constants. We all discussed that it seems reasonable to want to load a model even when it cannot be run, as there as the model object returned has lots of useful information that users may want (including, e.g., a list of constants, which they may not know before parsing the spec).

I'm here suggesting a mix:

- Use types that will give us static guarantees internally about when a model can be run (and maybe return this data also to users?)
- Allow initializing constants after load, but use a higher-level method rather than making users go through `assertConstraint`. I opted for this given that the `TransitionExecutor` also has an `initializeConstants` method, and it seems unkind to not give our users the same helper methods we give our programmers.

But I'm of course open to changes here.

Reply:
Yes, this makes sense. I have checked what `TransitionExecutorImpl.initializeConstants` is doing. Not only it uploads an assertion, but it also checks that all constants have been assigned symbolic values. However, `TransitionExecutorImpl.initializeConstants` accepts a single constraint, not a map from names to `TlaEx`. The reason is that we could add additional constraints in the expression. These additional constraints could be posed with `assert` too.

Reply:
So I guess the question is whether we want the user to have to form a constraint (using, what, and EDSL that constructs TLA+ expression? Just strings that the server will parse? I'm still not clear on that bit.) in order to initialize their constants, or if we let me do the high-level of thing of just setting these particular constraint values, and then they can use `assert` if they want to add additional constraints. 

I honestly don't know which users would prefer, by my general sense is that it's nice to make simple things simple, and support more advanced usage, but not require users to do advanced things to accomplish the simple ones.

Reply:
Yeah. I am also for keeping things simple. TLA+ specs have enough complexity already